### PR TITLE
[release/2.1] Do not add Request-Id header when its added already

### PR DESCRIPTION
**Pull request by lmolkova**
[edited by @danmosemsft ]

#### Description
This change prevents `Request-Id` header from being added by the [DiagnosticsHandler](https://github.com/dotnet/corefx/blob/master/src/System.Net.Http/src/System/Net/Http/DiagnosticsHandler.cs#L82) if it was already injected by tracing system.

This change allows a tracing system to implement backward compatibility with legacy correlation protocol after the new correlation protocol will be implemented in DiagnosticSource (3.0 timeframe).

See https://github.com/dotnet/corefx/issues/34299 for more details.

#### Customer Impact
Without this fix, after new DiagnosticSource is shipped, customers using new DiagnosticSource with 2.1 System.Net.Http AND forcing new [W3C distributed tracing protocol](https://github.com/w3c/distributed-tracing),  will not be able to correlate telemetry with other services that support legacy correlation protocol.
There will also be no workaround  for this.

#### Regression?
No.

#### Risk
Low.  @vancem : Note that the change is VERY small and safe (it is literally one line of code, that affects only one HTTP header (Request-id) and only if App-Insights did something (added one to override it)).   Thus it is very safe.  

#### Other options considered
* The Request-Id is current ‘standard’ for correlation supported by HttpClient, AspNetCore and Azure Application Insights, so using something else would have no effect on downstream services that expect Request-Id.
* We’ve discussed with Vance keeping DiagnosticSource in 3.0 backward compatible with HttpClient 2.1 and  after careful consideration decided that it would be a bad design choice, it would require additional public API surface changes.

### Comments from @vancem:
 Fundamentally.   App-Insights just wants its correlation features to work.   The thing that makes this problematic is that it needs the framework (in particular the HTTP subsystem), to help out (passing information in its headers), and sadly, there is non-trivial amount of industry-wide ‘churn’ on exactly how to do this.     Thus it requires changes to the framework if this churn is large enough (which it is at this time).    While many of the changes are covered by DiagnosticSource (which is as you point out of of band), some of the changes are in the HTTP logic, and thus part of the runtime.     Thus changes in how correlation information gets propagated can require runtime changes.        Moreover correlation by its very nature is about ‘big’ systems where it is likely that different parts evolve at different rates, and the requirement that everyone upgrade their runtime is likely to be an adoption-blocker.   
Looking at https://dotnet.microsoft.com/platform/support-policy/dotnet-core our only LTS runtime is 2.1.   Thus asking users to upgrade to the latest servicing of 2.1 is reasonable, at the present time it is unreasonable to ask them to upgrade to something beyond that (they lose their LTS).    
The alternative is to indicate that their new correlation work will not work until users upgrade to 3.0 (but frankly some would not do that until some post 3.0 build is declared LTS, which is quite a while from now).   
This is the driving force behind the desire for having the change in 2.1.
Just FYI,  Liudmila, Sergey and I met today and convinced me that with high likelihood they could achieve their goals with just this fix in V2.1 (the should not need another 2.1 fix in the future).   While there is always the possibility that we overlooked something, we did do some due diligence.
Dan: if you need me to offer more justification to Shiproom for this change let me know, I am solidly behind it.

<details>
  <summary>Original template</summary>
#### Description
This change prevents `Request-Id` header from being added by the [DiagnosticsHandler](https://github.com/dotnet/corefx/blob/master/src/System.Net.Http/src/System/Net/Http/DiagnosticsHandler.cs#L82) if it was already injected by tracing system.

This change allows a tracing system to implement backward compatibility with legacy correlation protocol after the new correlation protocol will be implemented in DiagnosticSource (3.0 timeframe).

See https://github.com/dotnet/corefx/issues/34299 for more details.

#### Customer Impact
Without this fix, after new DiagnosticSource is shipped, customers using new DiagnosticSource with 2.1 System.Net.Http AND forcing new [W3C distributed tracing protocol](https://github.com/w3c/distributed-tracing),  will not be able to correlate telemetry with other services that support legacy correlation protocol.
There will also be no workaround  for this.

#### Regression?
No.

#### Risk
Low. 

</details>

### CODE REVIEW

File: src/System.Net.Http/tests/FunctionalTests/DiagnosticsTests.cs [hunk 1]
```diff
@@ -381,6 +381,57 @@ public void SendAsync_ExpectedDiagnosticSourceActivityLogging()
             }, UseSocketsHttpHandler.ToString()).Dispose();
         }
 
+        [OuterLoop] // TODO: Issue #11345
```
Comment (Wraith2) [inline src/System.Net.Http/tests/FunctionalTests/DiagnosticsTests.cs]: This is the same comment as the test below, if the issue isn't relevant to this code can it be removed? from both?

**Comment by stephentoub** [comment]
Is this change already in master?

**Comment by lmolkova** [comment]
@stephentoub 
> Is this change already in master?

No, in master (3.0) we'll make another change. Does it make sense to push this change to master even though it will be changed again later?

**Comment by stephentoub** [comment]
> No, in master (3.0) we'll make another change. Does it make sense to push this change to master even though it will be changed again later?

I'll let @danmosemsft comment from a Shiproom perspective, but generally we port changes back from master rather than making specific changes in release branches; obviously there can be exceptions to that if it really doesn't make sense.  That said, the bar for release/2.1 is really high at this point.

**Comment by lmolkova** [comment]
here is PR against master https://github.com/dotnet/corefx/pull/34321

**Comment by lmolkova** [comment]
@vancem this is the only change that we need to the best of my knowledge.
I also want to port this change to 2.2

**Comment by lmolkova** [comment]
@dotnet-bot test this please

**Comment by danmosemsft** [comment]
> I also want to port this change to 2.2

We do not need a PR for that. Changes approved and merged into 2.1 get auto merged into 2.2.

File: src/System.Net.Http/tests/FunctionalTests/DiagnosticsTests.cs [hunk 2]
```diff
@@ -43,7 +43,7 @@ public static void EventSource_ExistsWithCorrectId()
         /// This test must be in the same test collection as any others testing HttpClient/WinHttpHandler
         /// DiagnosticSources, since the global logging mechanism makes them conflict inherently.
         /// </remarks>
-        [OuterLoop] // TODO: Issue #11345
+        [OuterLoop("Uses external server")]
```
Comment (danmosemsft) [inline src/System.Net.Http/tests/FunctionalTests/DiagnosticsTests.cs]: probably doesn't matter now you have the PR up but generally for backports we would not include unnecessary changes, so the diff is more obvious.

**Comment by lmolkova** [comment]
Build failure: 
 System.Net.Security.Tests.ServerAsyncAuthenticateTest fail with  `System.ComponentModel.Win32Exception : The client and server cannot communicate, because they do not possess a common algorithm`. This is not related to this PR.

there are a lot of closed issues related to these tests and perhaps there were some fixes that didn't make it into the 2.1. Also, other PRs against 2.1 fail with the same issues.  

Does it make sense to create an issue for these failures?

**Comment by danmosemsft** [comment]
@lmolkova most likely they are flaky, I would rerun the legs. If they are just flaky we might not do anything as there are so few checkins to 2.1.

@dotnet-bot test Windows x64 Debug Build  please
@dotnet-bot test Windows x64 Release Build  please

**Comment by danmosemsft** [comment]
Branch should be open tomorrow.

**Comment by danmosemsft** [comment]
2 System.Net.Security tests failed. @davidsh any context on this? Note this is the 2.1 branch. The change here *IS* to networking code, but my guess is that these are unrelated failures.
https://mc.dot.net/#/user/lmolkova/pr~2Fjenkins~2Fdotnet~2Fcorefx~2Frelease~2F2.1~2F/test~2Ffunctional~2Fcli~2F/8d71cae94239637c8a415662b416931d72acc332/workItem/System.Net.Security.Tests/analysis/xunit/System.Net.Security.Tests.ServerAsyncAuthenticateTest~2FServerAsyncAuthenticate_EachSupportedProtocol_Success(protocol:%20Ssl3)

```
Get Repro environment
Unhandled Exception of Type System.ComponentModel.Win32Exception
Message :
System.ComponentModel.Win32Exception : The client and server cannot communicate, because they do not possess a common algorithm
Stack Trace :
   at System.Net.SSPIWrapper.AcquireCredentialsHandle(SSPIInterface secModule, String package, CredentialUse intent, SCHANNEL_CRED scc) in D:\j\workspace\windows-TGrou---2c40d8a1\src\Common\src\Interop\Windows\sspicli\SSPIWrapper.cs:line 167
   at System.Net.Security.SslStreamPal.AcquireCredentialsHandle(CredentialUse credUsage, SCHANNEL_CRED secureCredential) in D:\j\workspace\windows-TGrou---2c40d8a1\src\System.Net.Security\src\System\Net\Security\SslStreamPal.Windows.cs:line 472
   at System.Net.Security.SslStreamPal.AcquireCredentialsHandle(X509Certificate certificate, SslProtocols protocols, EncryptionPolicy policy, Boolean isServer) in D:\j\workspace\windows-TGrou---2c40d8a1\src\System.Net.Security\src\System\Net\Security\SslStreamPal.Windows.cs:line 177
   at System.Net.Security.SecureChannel.AcquireClientCredentials(Byte[]& thumbPrint) in D:\j\workspace\windows-TGrou---2c40d8a1\src\System.Net.Security\src\System\Net\Security\SecureChannel.cs:line 603
   at System.Net.Security.SecureChannel.GenerateToken(Byte[] input, Int32 offset, Int32 count, Byte[]& output) in D:\j\workspace\windows-TGrou---2c40d8a1\src\System.Net.Security\src\System\Net\Security\SecureChannel.cs:line 802
   at System.Net.Security.SecureChannel.NextMessage(Byte[] incoming, Int32 offset, Int32 count) in D:\j\workspace\windows-TGrou---2c40d8a1\src\System.Net.Security\src\System\Net\Security\SecureChannel.cs:line 725
   at System.Net.Security.SslState.StartSendBlob(Byte[] incoming, Int32 count, AsyncProtocolRequest asyncRequest) in D:\j\workspace\windows-TGrou---2c40d8a1\src\System.Net.Security\src\System\Net\Security\SslState.cs:line 744
   at System.Net.Security.SslState.ForceAuthentication(Boolean receiveFirst, Byte[] buffer, AsyncProtocolRequest asyncRequest) in D:\j\workspace\windows-TGrou---2c40d8a1\src\System.Net.Security\src\System\Net\Security\SslState.cs:line 661
   at System.Net.Security.SslState.ProcessAuthentication(LazyAsyncResult lazyResult) in D:\j\workspace\windows-TGrou---2c40d8a1\src\System.Net.Security\src\System\Net\Security\SslState.cs:line 570
   at System.Net.Security.SslStream.BeginAuthenticateAsClient(SslClientAuthenticationOptions sslClientAuthenticationOptions, CancellationToken cancellationToken, AsyncCallback asyncCallback, Object asyncState) in D:\j\workspace\windows-TGrou---2c40d8a1\src\System.Net.Security\src\System\Net\Security\SslStream.cs:line 215
   at System.Net.Security.SslStream.BeginAuthenticateAsClient(String targetHost, X509CertificateCollection clientCertificates, SslProtocols enabledSslProtocols, Boolean checkCertificateRevocation, AsyncCallback asyncCallback, Object asyncState) in D:\j\workspace\windows-TGrou---2c40d8a1\src\System.Net.Security\src\System\Net\Security\SslStream.cs:line 195
   at System.Net.Security.SslStream.<>c.<AuthenticateAsClientAsync>b__46_1(String arg1, X509CertificateCollection arg2, SslProtocols arg3, AsyncCallback callback, Object state) in D:\j\workspace\windows-TGrou---2c40d8a1\src\System.Net.Security\src\System\Net\Security\SslStream.cs:line 383
   at System.Threading.Tasks.TaskFactory`1.FromAsyncImpl[TArg1,TArg2,TArg3](Func`6 beginMethod, Func`2 endFunction, Action`1 endAction, TArg1 arg1, TArg2 arg2, TArg3 arg3, Object state, TaskCreationOptions creationOptions)
   at System.Net.Security.SslStream.AuthenticateAsClientAsync(String targetHost, X509CertificateCollection clientCertificates, SslProtocols enabledSslProtocols, Boolean checkCertificateRevocation) in D:\j\workspace\windows-TGrou---2c40d8a1\src\System.Net.Security\src\System\Net\Security\SslStream.cs:line 384
   at System.Net.Security.Tests.ServerAsyncAuthenticateTest.ServerAsyncSslHelper(SslProtocols clientSslProtocols, SslProtocols serverSslProtocols, Boolean expectedToFail) in D:\j\workspace\windows-TGrou---2c40d8a1\src\System.Net.Security\tests\FunctionalTests\ServerAsyncAuthenticateTest.cs:line 127
   at System.Net.Security.Tests.ServerAsyncAuthenticateTest.ServerAsyncAuthenticate_EachSupportedProtocol_Success(SslProtocols protocol) in D:\j\workspace\windows-TGrou---2c40d8a1\src\System.Net.Security\tests\FunctionalTests\ServerAsyncAuthenticateTest.cs:line 42
--- End of stack trace from previous location where exception was thrown ---
```
@dotnet-bot test Windows x64 Debug Build please
@dotnet-bot test Windows x64 Release Build please
@dotnet-bot test Windows x86 Debug Build please
@dotnet-bot test Windows x86 Release Build please

**Comment by davidsh** [comment]
>System.Net.Security tests failed. @davidsh any context on this? Note this is the 2.1 branch. The change here IS to networking code, but my guess is that these are unrelated failures.

Probably not related.  See #34424 for similar test failures.

**Comment by jamshedd** [comment]
Approved for 2.1.8.

**Comment by jamshedd** [comment]
Approved for 2.1.8.

**Comment by vancem** [comment]
Also we have failures but given the discussion above I am assuming that is OK, right?

**Comment by danmosemsft** [comment]
Test failures are unrelated as mentioned above.
